To answer this question, we need to extract specific information from the `features` table for each transcript isoform of the specified genes (XIST, MALAT1, BRCA1, NFKB1, COL1A2, HOTAIR, HOXC11, and HOXC12). We will also use the `relations` table to determine which features are exons of a given transcript.

Here's how we can approach this:

### Step 1: Identify Transcripts for Each Gene
First, we need to find all transcripts (features with type "transcript") associated with each gene. The genes' names will be used as attributes in the `attributes` column of the `features` table.

```sql
SELECT id AS transcript_id, seqid AS chromosomal_location, start AS position_start, end AS position_end, strand 
FROM features 
WHERE featuretype = 'transcript' AND attributes LIKE '%gene_name "%XIST%"%' OR attributes LIKE '%gene_name "%MALAT1%"%' OR attributes LIKE '%gene_name "%BRCA1%"%' OR attributes LIKE '%gene_name "%NFKB1%"%' OR attributes LIKE '%gene_name "%COL1A2%"%' OR attributes LIKE '%gene_name "%HOTAIR%"%' OR attributes LIKE '%gene_name "%HOXC11%"%' OR attributes LIKE '%gene_name "%HOXC12%"%';
```

### Step 2: Count the Number of Exons for Each Transcript
Next, we need to count how many exons each transcript has. We can do this by joining the `features` table with itself on the parent-child relationship stored in the `relations` table.

```sql
SELECT t1.id AS transcript_id, COUNT(t2.id) AS number_of_exons 
FROM features AS t1 
JOIN relations AS r ON t1.id = r.parent 
JOIN features AS t2 ON r.child = t2.id AND t2.featuretype = 'exon' 
WHERE t1.featuretype = 'transcript' AND (t1.attributes LIKE '%gene_name "%XIST%"%' OR t1.attributes LIKE '%gene_name "%MALAT1%"%' OR t1.attributes LIKE '%gene_name "%BRCA1%"%' OR t1.attributes LIKE '%gene_name "%NFKB1%"%' OR t1.attributes LIKE '%gene_name "%COL1A2%"%' OR t1.attributes LIKE '%gene_name "%HOTAIR%"%' OR t1.attributes LIKE '%gene_name "%HOXC11%"%' OR t1.attributes LIKE '%gene_name "%HOXC12%"%');
```

### Step 3: Calculate the Average Exon Size for Each Transcript
To calculate the average exon size, we need to sum up the sizes of all exons and divide by the number of exons.

```sql
SELECT r.parent AS transcript_id, SUM(t2.end - t2.start + 1) / COUNT(*) AS avg_exon_size 
FROM relations AS r 
JOIN features AS t2 ON r.child = t2.id AND t2.featuretype = 'exon' 
WHERE r.parent IN (
    SELECT id FROM features WHERE featuretype = 'transcript' AND (attributes LIKE '%gene_name "%XIST%"%' OR attributes LIKE '%gene_name "%MALAT1%"%' OR attributes LIKE '%gene_name "%BRCA1%"%' OR attributes LIKE '%gene_name "%NFKB1%"%' OR attributes LIKE '%gene_name "%COL1A2%"%' OR attributes LIKE '%gene_name "%HOTAIR%"%' OR attributes LIKE '%gene_name "%HOXC11%"%' OR attributes LIKE '%gene_name "%HOXC12%"%')
) 
GROUP BY r.parent;
```

### Step 4: Combine All Information
Finally, we need to combine all the information into a single table. We will join the results from the previous queries and add the biotype (which is part of the `attributes` column).

```sql
WITH transcripts AS (
    SELECT id AS transcript_id, seqid AS chromosomal_location, start AS position_start, end AS position_end, strand 
    FROM features 
    WHERE featuretype = 'transcript' AND attributes LIKE '%gene_name "%XIST%"%' OR attributes LIKE '%gene_name "%MALAT1%"%' OR attributes LIKE '%gene_name "%BRCA1%"%' OR attributes LIKE '%gene_name "%NFKB1%"%' OR attributes LIKE '%gene_name "%COL1A2%"%' OR attributes LIKE '%gene_name "%HOTAIR%"%' OR attributes LIKE '%gene_name "%HOXC11%"%' OR attributes LIKE '%gene_name "%HOXC12%"%'
),
exon_counts AS (
    SELECT t1.id AS transcript_id, COUNT(t2.id) AS number_of_exons 
    FROM features AS t1 
    JOIN relations AS r ON t1.id = r.parent 
    JOIN features AS t2 ON r.child = t2.id AND t2.featuretype = 'exon' 
    WHERE t1.featuretype = 'transcript'
),
avg_exon_sizes AS (
    SELECT r.parent AS transcript_id, SUM(t2.end - t2.start + 1) / COUNT(*) AS avg_exon_size 
    FROM relations AS r 
    JOIN features AS t2 ON r.child = t2.id AND t2.featuretype = 'exon' 
    WHERE r.parent IN (SELECT id FROM transcripts)
    GROUP BY r.parent
),
biotypes AS (
    SELECT id AS transcript_id, SUBSTR(attributes, INSTR(attributes, 'transcript_biotype "') + 18, INSTR(attributes, '"', INSTR(attributes, 'transcript_biotype "')) - INSTR(attributes, 'transcript_biotype "') - 18) AS biotype 
    FROM features 
    WHERE featuretype = 'transcript'
)
SELECT t.transcript_id, t.chromosomal_location, t.position_start, t.position_end, e.number_of_exons, a.avg_exon_size, t.strand, b.biotype
FROM transcripts t
JOIN exon_counts e ON t.transcript_id = e.transcript_id
JOIN avg_exon_sizes a ON t.transcript_id = a.transcript_id
JOIN biotypes b ON t.transcript_id = b.transcript_id;
```

This query will give you the required information about each transcript isoform of the specified genes, organized in a table format. Make sure to adjust the `attributes` column parsing logic if your GFF/GTF file uses different formats for storing gene names and biotypes.